src/Analyzers/AnalyzerFormatter.cs

```diff
@@ -33,59 +36,85 @@ internal class AnalyzerFormatter : ICodeFormatter
-            }
-            else
-            {
-                solution = await FixDiagnosticsAsync(solution, formattableDocuments, logger, cancellationToken);
+                return solution;
             }
 
+            var analysisStopwatch = Stopwatch.StartNew();
+            logger.LogTrace($"Running {_name} analysis.");
+
+            var formattablePaths = formattableDocuments.Select(id => solution.GetDocument(id)?.FilePath)
+                    .OfType<string>().ToImmutableHashSet();
+
+            logger.LogTrace("Determining diagnostics...");
+
+            var allAnalyzers = analyzersAndFixers.Select(pair => pair.Analyzer).ToImmutableArray();
+            var projectAnalyzers = await _finder.FilterBySeverityAsync(solution.Projects, allAnalyzers, formattablePaths, formatOptions, cancellationToken).ConfigureAwait(false);
+
+            var projectDiagnostics = await GetProjectDiagnosticsAsync(solution, projectAnalyzers, formattablePaths, formatOptions, logger, formattedFiles, cancellationToken).ConfigureAwait(false);
+
+            var projectDiagnosticsMS = analysisStopwatch.ElapsedMilliseconds;
+            logger.LogTrace(Resources.Complete_in_0_ms, projectDiagnosticsMS);
+
+            logger.LogTrace("Fixing diagnostics...");
+
+            solution = await FixDiagnosticsAsync(solution, analyzersAndFixers, projectDiagnostics, formattablePaths, logger, cancellationToken).ConfigureAwait(false);
+
+            var fixDiagnosticsMS = analysisStopwatch.ElapsedMilliseconds - projectDiagnosticsMS;
+            logger.LogTrace(Resources.Complete_in_0_ms, fixDiagnosticsMS);
+
             logger.LogTrace("Analysis complete in {0}ms.", analysisStopwatch.ElapsedMilliseconds);
 
             return solution;
         }
 
-        private async Task LogDiagnosticsAsync(Solution solution, ImmutableArray<DocumentId> formattableDocuments, FormatOptions options, ILogger logger, CancellationToken cancellationToken)
+        private async Task<ImmutableDictionary<ProjectId, ImmutableHashSet<string>>> GetProjectDiagnosticsAsync(
+            Solution solution,
+            ImmutableDictionary<Project, ImmutableArray<DiagnosticAnalyzer>> projectAnalyzers,
+            ImmutableHashSet<string> formattablePaths,
+            FormatOptions options,
+            ILogger logger,
+            List<FormattedFile> formattedFiles,
+            CancellationToken cancellationToken)
         {
-            var pairs = _finder.GetAnalyzersAndFixers();
-            var paths = formattableDocuments.Select(id => solution.GetDocument(id)?.FilePath)
-                .OfType<string>().ToImmutableArray();
-
-            // no need to run codefixes as we won't persist the changes
-            var analyzers = pairs.Select(x => x.Analyzer).ToImmutableArray();
             var result = new CodeAnalysisResult();
-            await solution.Projects.ForEachAsync(async (project, token) =>
+            foreach (var project in solution.Projects)
```

Comment:
Any reason to not kick all this work off in parallel?

Reply:
We parallelize the analyzers